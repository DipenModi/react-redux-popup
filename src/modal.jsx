import { Component, PropTypes } from 'react';
import HigherOrderPopupComponent from 'rrp/higher-order-popup-component';
import { TYPE_MODAL } from 'rrp/popup-collection';

const PROP_TYPES = {
    id: PropTypes.string.isRequired,
    layoverClassName: PropTypes.string,
    popupClassName: PropTypes.string,
    style: PropTypes.object
};

export default function(ComposedComponent) {
    class Modal extends Component {
        constructor(props) {
            super(props);
            this.state = { style: {} };
            this.setModalRef = el => {
                this.modal = el;
            };
        }

        componentDidMount() {
            const { clientHeight, clientWidth } = this.modal;
            // eslint-disable-next-line react/no-did-mount-set-state
            this.setState({
                style: {
                    left: (window.innerWidth - clientWidth) / 2,
                    top: (window.innerHeight - clientHeight) / 2
                }
            });
        }

        render() {
            const className = `${this.props.popupClassName ? this.props.popupClassName : ''}`;
            const style = {
                ...this.state.style,
                ...this.props.style
            };

            return (
                <div id={this.props.id}>
                    <div className={this.props.layoverClassName} />
                    <div className={className} ref={this.setModalRef} style={style}>
                        <ComposedComponent {...this.props} />
                    </div>
                </div>
            );
        }
    }

    Modal.propTypes = PROP_TYPES;

    return HigherOrderPopupComponent(Modal, TYPE_MODAL);
}
