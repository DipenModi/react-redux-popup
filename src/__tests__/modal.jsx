import { shallow } from 'enzyme';
import { Modal } from 'rrp/modal';

describe('modal', () => {
    // eslint-disable-next-line no-console
    const origConsoleError = console.error;

    const renderComponent = () => <div>View</div>;

    afterEach(() => {
        // eslint-disable-next-line no-console
        console.error = origConsoleError;
    });

    it('should render the component', () => {
        const wrapper = shallow(
            <Modal
                id="modal1"
                render={renderComponent}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should render the render prop', () => {
        const wrapper = shallow(
            <Modal
                anchor="left"
                className="my-popup"
                id="modal1"
                layoverClassName="my-layover"
                render={renderComponent}
                style={{ color: '#fff' }}
                transitionEnterTimeout={300}
                transitionExitTimeout={300}
                transitionName="modal"
            />
        );
        const childWrapper = shallow(wrapper.prop('render')(), { disableLifecycleMethods: true });
        expect(childWrapper).toMatchSnapshot();
    });

    it('should render the render prop', () => {
        const wrapper = shallow(
            <Modal
                anchor="left"
                className="my-popup"
                id="modal1"
                layoverClassName="my-layover"
                render={renderComponent}
                style={{ color: '#fff' }}
                transitionEnterTimeout={300}
                transitionExitTimeout={300}
                transitionName="modal"
            />
        );
        const childWrapper = shallow(wrapper.prop('render')(), { disableLifecycleMethods: true });
        expect(childWrapper).toMatchSnapshot();
    });

    it('should render with modal prop', () => {
        const wrapper = shallow(
            <Modal
                anchor="left"
                className="my-popup"
                id="modal1"
                layoverClassName="my-layover"
                modal1
                render={renderComponent}
                transitionEnterTimeout={300}
                transitionExitTimeout={300}
                transitionName="modal"
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should warn about missing id', () => {
        const mockConsole = jest.fn();
        // eslint-disable-next-line no-console
        console.error = mockConsole;

        shallow(<Modal render={renderComponent} />);
        expect(mockConsole).toHaveBeenCalledTimes(1);
        expect(mockConsole).toHaveBeenCalledWith(
            expect.stringContaining('The prop `id` is marked as required'));
    });

    it('should warn about missing render prop', () => {
        const mockConsole = jest.fn();
        // eslint-disable-next-line no-console
        console.error = mockConsole;

        shallow(<Modal id="modal1" />);
        expect(mockConsole).toHaveBeenCalledTimes(1);
        expect(mockConsole).toHaveBeenCalledWith(
            expect.stringContaining('The prop `render` is marked as required'));
    });
});
