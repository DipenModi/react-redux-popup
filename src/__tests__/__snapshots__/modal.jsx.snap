// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal should render the component 1`] = `
<TransitionWrapper
  getPortalElement={[Function]}
  isOpen={false}
  isPortalReady={false}
  render={[Function]}
  transitionEnterTimeout={300}
  transitionExitTimeout={300}
  transitionName="modal"
/>
`;

exports[`modal should render the render prop 1`] = `
<div>
  <div
    className="my-layover"
  />
  <div
    className="my-popup"
    style={
      Object {
        "color": "#fff",
        "left": 0,
        "top": 0,
      }
    }
  >
    <div>
      View
    </div>
  </div>
</div>
`;

exports[`modal should render the render prop 2`] = `
<div>
  <div
    className="my-layover"
  />
  <div
    className="my-popup"
    style={
      Object {
        "color": "#fff",
        "left": 0,
        "top": 0,
      }
    }
  >
    <div>
      View
    </div>
  </div>
</div>
`;

exports[`modal should render with modal prop 1`] = `
<TransitionWrapper
  getPortalElement={[Function]}
  isOpen={true}
  isPortalReady={false}
  render={[Function]}
  transitionEnterTimeout={300}
  transitionExitTimeout={300}
  transitionName="modal"
/>
`;
